runtime/ast/programindices.go

```diff
@@ -0,0 +1,112 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright 2019-2020 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ast
+
+import (
+	"sync"
+)
+
+// programIndices is a container for all indices of a program's declarations
+//
+type programIndices struct {
+	once sync.Once
+	// Use `pragmaDeclarations` instead
+	_pragmaDeclarations []*PragmaDeclaration
+	// Use `importDeclarations` instead
+	_importDeclarations []*ImportDeclaration
+	// Use `interfaceDeclarations` instead
+	_interfaceDeclarations []*InterfaceDeclaration
+	// Use `compositeDeclarations` instead
+	_compositeDeclarations []*CompositeDeclaration
+	// Use `functionDeclarations()` instead
+	_functionDeclarations []*FunctionDeclaration
+	// Use `transactionDeclarations()` instead
+	_transactionDeclarations []*TransactionDeclaration
+}
+
+func (i *programIndices) pragmaDeclarations(declarations []Declaration) []*PragmaDeclaration {
```

Comment:
We are assuming here that these methods are always called with the same `[]Declarations`, otherwise the declarations cached by the `initializer` function will be inconsistent, which makes me a bit nervous. This method is called from the `Program` method like `p.indices.transactionDeclarations(p.Declarations)`, and `Declarations` is a public field of `Program`. Any code with access to a `Program` instance could change `Declarations` and make all these methods incorrect.

Reply:
This is correct, and I don't like it, though I'm not sure how to improve this in Go.
Basically all I want is lazy fields (for the indices) and having the `Declarations` field immutable. If you have any ideas, please let me know.

Reply:
👍 Maybe making `Declarations` private and exposing a getter publicly would help? At least then it's not as easy to modify the field outside the package.